cvmfs/util.h

```diff
@@ -279,32 +279,66 @@ struct hash_murmur {
 std::string Base64(const std::string &data);
 bool Debase64(const std::string &data, std::string *decoded);
 
-template <class T>
-class UniquePtr : SingleCopy {
+template <class T, class DerivedT>
+class UniquePtrBase : SingleCopy {
  public:
-  inline UniquePtr() : ref_(NULL) {}
-  inline explicit UniquePtr(T *ref) : ref_(ref) { }
-  inline ~UniquePtr()                 { delete ref_; }
+  inline UniquePtrBase() : ref_(NULL) {}
+  inline explicit UniquePtrBase(T *ref) : ref_(ref) { }
+  inline ~UniquePtrBase()                 { Free(); }
 
   inline operator bool() const        { return IsValid(); }
-  inline T& operator*() const         { return *ref_; }
   inline T* operator->() const        { return ref_; }
-  inline UniquePtr& operator=(T* ref) {
+  inline operator T*()                { return ref_; }
+  inline DerivedT& operator=(T *ref) {
     if (ref_ != ref) {
-      delete ref_;
+      Free();
       ref_ = ref;
     }
-    return *this;
+    return *(static_cast<DerivedT*>(this));
   }
-
   inline T* weak_ref() const          { return ref_; }
   inline bool IsValid() const         { return (ref_ != NULL); }
   inline T*   Release()               { T* r = ref_; ref_ = NULL; return r; }
 
- private:
+ protected:
+  void Free() {
+    static_cast<DerivedT*>(this)->Free();
+  }
   T *ref_;
 };
 
+template <class T>
+class UniquePtr : public UniquePtrBase<T, UniquePtr<T> > {
+ private:
+  typedef UniquePtrBase<T, UniquePtr<T> > BaseT;
+ public:
+  friend class UniquePtrBase<T, UniquePtr<T> >;
+  using BaseT::operator=;
```

Comment:
Concretely the _operator=_ is by default both overwritten and inherited. The first one implies that the second one is _hidden_, so that _ussing BaseT::operator=_ tells the compiler not to try to create another operator= for the derived class which would hide the inherited one.